### DISCUSSION TITLE
IPVLAN L3 Driver Bug or How Does It Work?

### DISCUSSION THREAD

**Discussion by josoko** [moby/moby#48051] [General]
Hello

I want to create 3 IPVLAN Mode L3 network instances on my Docker host:

```
172.16.101.16/28 => Name: network1
172.16.101.32/28 => Name: network2
172.16.101.45/28 => Name: network3
```

In order for the network instance to reach the GW and thus IP addresses outside the network instance, I have to specify the parent interface when creating the instance. Otherwise it won't work.

Example:

```
docker network create -d ipvlan \
--subnet 172.16.101.16/28 \
-o parent=eno1 -o ipvlan_mode=l3 \
network1
```

If I create this one network instance, everything works.

The problem is that I can't create another network instance. For example, this one:

```
docker network create -d ipvlan \
--subnet 172.16.101.32/28 \
-o parent=eno1 -o ipvlan_mode=l3 \
network2
```

When I try to create the second network instance, I get this error message:

```
Error response from daemon: network di-0be9895ca25b is already using parent interface eno1
```

What would work would be this:

```
docker network create -d ipvlan \
--subnet 172.16.101.16/28 \
-o parent=eno1 -o ipvlan_mode=l3 \
--subnet 172.16.101.32/28 \
--subnet 172.16.101.48/28 \
network1
```

But then everything runs under the same network instance and I can't separate the networks by apps. If it doesn't work on the CLI, then it won't work in Docker Compose files either. But it should actually work. Because you can create different network instances if it's not IPVLAN L3.

What I would like to do is:
- Docker Compose file docker-compose-app1.yml will use the network1 for its hosts.
- Docker Compose file docker-compose-app2.yml will use the network2 for its hosts.

Does anyone have experience with it?

**Comment by thaJeztah** [comment]
Seems related to https://github.com/moby/moby/pull/47318#issuecomment-2169540904

cc @robmry @akerouanton

**Reply by robmry** [reply]
Yes, looks like ipvlan would need a change like macvlan's.

**Reply by josoko** [reply]
Addendum:

The current Docker version I am using is:

```
docker --version 
Docker version 26.1.4, build 5650f9b
```

I am pretty confident that this feature improvement will be implement by you.

How can I track the this case to know in which version this feature will be implemented?

**Comment by josoko** [comment]
@thaJeztah 
Sorry, again me. 

Seems it has not been implemented in 27.0.1 or 27.0.2.
Do you have this topic on the radar?

**Reply by thaJeztah** [reply]
I don't think anyone has time yet to look into that part, but contributions welcome if you're interested 👍

**Comment by OOP-778** [comment]
Any updates on this issue?

**Reply by robmry** [reply]
Hi @OOP-778 - I don't think anyone's looked at it.

So, if you're asking because you're interested in contributing (?), that'd be great!

**Comment by marcinosypka** [comment]
I would love to see this implemented.
